services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/internal/ProjectionExpression.java

```diff
@@ -0,0 +1,170 @@
+ * contains the mapping from the placeholder attribute name to the actual attribute name (see
+ * {@link #expressionAttributeNames()}).
+ * <p>
+ * <b>Resolving duplicates</b>
+ * <ul>
+ *     <li>If the input to the ProjectionExpression contains the same attribute name in more than one place, independent of
+ *     nesting level, it will be mapped to a single placeholder</li>
+ *     <li>If two attributes resolves to the same placeholder name, a disambiguator is added to the placeholder in order to
+ *     make it unique.</li>
+ * </ul>
+ * <p>
+ * <b>Placeholder conversion examples</b>
+ * <ul>
+ *     <li>'MyAttribute' maps to {@code #AMZN_MAPPED_MyAttribute}</li>
+ *     <li>'MyAttribute' appears twice in input but maps to only one entry {@code #AMZN_MAPPED_MyAttribute}.</li>
+ *     <li>'MyAttribute-1' maps to {@code #AMZN_MAPPED_MyAttribute_1}</li>
+ *     <li>'MyAttribute-1' and 'MyAttribute.1' in the same input maps to {@code #AMZN_MAPPED_0_MyAttribute_1} and
+ *     {@code #AMZN_MAPPED_1_MyAttribute_1}</li>
+ * </ul>
+ * <b>Projection expression usage example</b>
+ * <pre>
+ * {@code
+ * List<NestedAttributeName> attributeNames = Arrays.asList(
+ *     NestedAttributeName.create("MyAttribute")
+ *     NestedAttributeName.create("MyAttribute.WithDot", "MyAttribute.03"),
+ *     NestedAttributeName.create("MyAttribute:03, "MyAttribute")
+ * );
+ * ProjectionExpression projectionExpression = ProjectionExpression.create(attributeNames);
+ * Map<String, String> expressionAttributeNames = projectionExpression.expressionAttributeNames();
+ * Optional<String> projectionExpressionString = projectionExpression.projectionExpressionAsString();
+ * }
+ *
+ * results in
+ *
+ * expressionAttributeNames: {
+ *    #AMZN_MAPPED_MyAttribute : MyAttribute,
+ *    #AMZN_MAPPED_MyAttribute_WithDot : MyAttribute.WithDot}
+ *    #AMZN_MAPPED_0_MyAttribute_03 : MyAttribute.03}
+ *    #AMZN_MAPPED_1_MyAttribute_03 : MyAttribute:03}
+ * }
+ * and
+ *
+ * projectionExpressionString: "#AMZN_MAPPED_MyAttribute,#AMZN_MAPPED_MyAttribute_WithDot.#AMZN_MAPPED_0_MyAttribute_03,
+ *                              #AMZN_MAPPED_1_MyAttribute_03.#AMZN_MAPPED_MyAttribute"
+ * </pre>
+ * <p>
+ * For more information, see <a href=
+ * "https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ProjectionExpressions.html"
+ * >Projection Expressions</a> in the <i>Amazon DynamoDB Developer Guide</i>.
+ * </p>
+ */
+@SdkInternalApi
+public class ProjectionExpression {
+
+    private static final String AMZN_MAPPED = "#AMZN_MAPPED_";
+    private static final UnaryOperator<String> PROJECTION_EXPRESSION_KEY_MAPPER = k -> AMZN_MAPPED + cleanAttributeName(k);
+
+    private final Optional<String> projectionExpression;
+    private final Map<String, String> expressionAttributeNames;
+
+    private ProjectionExpression(List<NestedAttributeName> nestedAttributeNames) {
+        this.expressionAttributeNames = createAttributePlaceholders(nestedAttributeNames);
```

Comment:
Should we create a unmodifable copy of the map?

Reply:
Updated